### DISCUSSION TITLE
How to get the sub-item text of the selected item in the list view in WPF?

### DISCUSSION THREAD

**Discussion by VahidEra** [dotnet/wpf#7430] [Q&A] [answered]
Hi guys,  
I have a list view in my WPF project that has two columns name and address and I want to get the text of a certain column by clicking on the desired item, for example the address column.  I used to do this easily in win form, but in WPF, unfortunately, I still haven't managed to do this.

**This is my cod in desing view:**
```
       <ListView PreviewMouseDown="lis_MouseDown" 
                  x:Name="lis" 
                  HorizontalAlignment="Left" 
                  Height="272" 
                  Margin="392,51,0,0" 
                  VerticalAlignment="Top"
                  Width="317" 
                  d:ItemsSource="{d:SampleData ItemCount=5}">
            <ListView.View>
                <GridView>
                    <GridViewColumn Header="name" DisplayMemberBinding="{Binding name}"/>
                    <GridViewColumn  Header="address" DisplayMemberBinding="{Binding address}"/>
                </GridView>
            </ListView.View>
        </ListView>
```

**and I tried this :**

```
Dim str As String = CType(CType(lis.SelectedItems(0), DataRowView)("address"), String)
lbltest.Content = str
```
**But unfortunately, I keep getting errors like this:**
>System.InvalidCastException: 'Unable to cast object of type 'VB$AnonymousType_0`2[System.String,System.String]' to type 'System.Data.DataRowView'.'

Can you help me with this? 
I just want to get the subitem, and you can also change my code.

**Comment by miloush** [comment]
Can you provide a project we can run? Your code doesn't show what the data source is.

**Reply by VahidEra** [reply]
@miloush Dear friend, of course.
This is a sample project:
https://1drv.ms/u/s!AqwLh1F7nIIAgxLpBGBI1xjxlq2i?e=xXB7we

**Comment by miloush** [comment, marked as answer]
Okay so your data, as the error already suggests, comes from an anonymous type:

```vb
        lis.Items.Add(New With {Key .name = "Jack", Key .address = "3240 Vernon Street"})
        lis.Items.Add(New With {Key .name = "Mary", Key .address = "3729 Cambridge Place"})
```

The lis.SelectedItem will contain the same objects you put in, not a "DataRowView". If you make up a type on the go as you did, it will be an anonymous type. You cannot access its members without Option Strict Off (or reflection).

A better option is to declare your type ahead of time:

```vb
Class MyRowData
    Public Property name As String
    Public Property address As String
End Class

Class MainWindow
    Private Sub MainWindow_Loaded(sender As Object, e As RoutedEventArgs) Handles Me.Loaded
        lis.Items.Add(New MyRowData With {.name = "Jack", .address = "3240 Vernon Street"})
        lis.Items.Add(New MyRowData With {.name = "Mary", .address = "3729 Cambridge Place"})
    End Sub

    Private Sub lis_PreviewMouseDoubleClick(sender As Object, e As MouseButtonEventArgs) Handles lis.PreviewMouseDoubleClick
        Dim str As String = CType(lis.SelectedItems(0), MyRowData).address
        lbltest.Content = str
    End Sub
End Class
```

**Reply by VahidEra** [reply]
@miloush wow, thank you very much dear friend. This was exactly what I wanted.  Thank you very much for your help.
